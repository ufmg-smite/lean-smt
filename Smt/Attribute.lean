import Lean
import Std

namespace Smt.Attribute

open Lean
open Std

/-- An extension to Lean's runtime environment to support SMT attributes.
    Maintains a set of function declarations for the `smt` tactic to utilize
    while generating the SMT query. -/
abbrev SmtExtension := SimpleScopedEnvExtension Name (HashSet Name)

/-- Adds a declaration to the set of function declarations maintained by the SMT
    environment extension. -/
def addSmtEntry (d : HashSet Name) (e : Name) : HashSet Name :=
  d.insert e

initialize smtExt : SmtExtension ← registerSimpleScopedEnvExtension {
  name     := `SmtExt
  initial  := {}
  addEntry := addSmtEntry
}

/-- Throws unexpected type error. -/
def throwUnexpectedType (t : Name) (n : Name) : AttrM Unit :=
  throwError s!"unexpected type at '{n}', `{t}` expected"

/-- Validates the tagged declaration. -/
def validate (n : Name) : AttrM Unit := do
  match (← getEnv).find? n with
  | none      => throwError s!"unknown constant '{n}'"
  | some info =>
    match info.type with
    | Expr.const c .. =>
      if c != (`Smt.Transformer) then throwUnexpectedType `Smt.Transformer n
    | _               => throwUnexpectedType `Smt.Transformer n

/-- Registers an SMT attribute with the provided name and description and links
    it against `ext`. -/
def registerSmtAttr (attrName : Name) (attrDescr : String)
  : IO Unit :=
  registerBuiltinAttribute {
    name  := attrName
    descr := attrDescr
    applicationTime := AttributeApplicationTime.afterTypeChecking
    add   := fun decl stx attrKind => do
      trace[Smt.debug.attr] s!"attrName: {attrName}, attrDescr: {attrDescr}"
      trace[Smt.debug.attr] s!"decl: {decl}, stx: {stx}, attrKind: {attrKind}"
      Attribute.Builtin.ensureNoArgs stx
      validate decl
      setEnv (smtExt.addEntry (← getEnv) decl)
      trace[Smt.debug.attr]
        s!"transformers: {(smtExt.getState (← getEnv)).toList}"
    erase := fun declName => do
      let s := smtExt.getState (← getEnv)
      let s := s.erase declName
      modifyEnv fun env => smtExt.modifyState env fun _ => s
  }

initialize registerSmtAttr `Smt
             "Utilize this function to transform Lean expressions to SMT terms."

end Smt.Attribute
