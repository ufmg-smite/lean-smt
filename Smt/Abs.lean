/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Translator
import Mathlib.Algebra.Order.Group.Abs

namespace Smt.Abs

open Lean Expr
open Translator Term


@[smtTranslator] def replaceAbs : Translator
  | app (app (app (app (const ``abs _) _) _) _) e => do
      let tmE ← applyTranslators! e
      return appT (symbolT "abs") tmE
  | _ => return none

end Smt.Abs
