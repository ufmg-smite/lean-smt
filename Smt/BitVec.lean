/-
Copyright (c) 2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Wojciech Nawrocki
-/

import Lean
import Smt.Translator
import Smt.Data.BitVec

namespace Smt.BitVec

open Lean Expr
open Translator Term

@[smtTranslator] def replaceType : Translator
  | e@(app (const ``BitVec _) n) => do
    let n ← Meta.whnf n
    let some n ← Meta.evalNat n |>.run
      | throwError "bitvector type{indentD e}\nhas variable width"
    return mkApp2 (symbolT "_") (symbolT "BitVec") (literalT (toString n))
  | _ => return none

def mkLit (w : Nat) (n : Nat) : Term :=
  let bits := Nat.toDigits 2 n |>.take w
  literalT <| bits.foldl (init := "#b".pushn '0' (w - bits.length)) (·.push ·)

@[smtTranslator] def replaceEq : Translator
  -- TODO: we should really just support beq/bne across all types uniformly
  | app (app (const ``BEq.beq _) (app (const ``BitVec _) _)) _ => return symbolT "="
  | app (app (app (app (const ``bne _) (app (const ``BitVec _) _)) _) e) e' =>
    return appT (symbolT "not") (mkApp2 (symbolT "=") (← applyTranslators! e) (← applyTranslators! e'))
  | _ => return none

@[smtTranslator] def replaceFun : Translator
  | app (app (const ``BitVec.append _) _) _ => return symbolT "concat"
  | app (const ``BitVec.and _) _            => return symbolT "bvand"
  | app (const ``BitVec.or _) _             => return symbolT "bvor"
  | app (const ``BitVec.xor _) _            => return symbolT "bvxor"
  | e@(app (app (app (const ``BitVec.shiftLeft _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvshl") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.shiftRight _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvlshr") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.extract _) _) i) j) => do
    let i ← reduceLit i e
    let j ← reduceLit j e
    return mkApp3 (symbolT "_") (symbolT "extract") (literalT (toString i)) (literalT (toString j))
  | e@(app (const ``BitVec.zero _) w) => do
    let w ← reduceWidth w e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w 0
  | e@(app (app (const ``BitVec.ofNat _) w) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w n
  | _ => return none
where
  reduceWidth (w : Expr) (e : Expr) : TranslationM Nat := do
    let some w ← Meta.evalNat (← Meta.whnf w) |>.run
      | throwError "bitvector width{indentD w}\nis not constant in{indentD e}"
    return w
  reduceLit (n : Expr) (e : Expr) : TranslationM Nat := do
    let some n ← Meta.evalNat (← Meta.whnf n) |>.run
      | throwError "literal{indentD n}\nis not constant in{indentD e}"
    return n

end Smt.BitVec
