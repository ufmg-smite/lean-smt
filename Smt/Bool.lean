/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Transformer

namespace Smt.Bool

open Lean Expr
open Smt.Transformer

@[Smt] def replaceConst : Transformer
  | const `Bool.true ..                                 => pure (mkConst `true)
  | const `Bool.false ..                                => pure (mkConst `false)
  | app (app (const `BEq.beq ..) (const `Bool ..) _) .. =>
    pure (mkConst (Name.mkSimple "="))
  | e                                                   => pure e

@[Smt] def replaceEq : Transformer
  | app (app (const `Decidable.decide ..)
             (app (app (app (const `Eq ..) (const `Bool ..) _)
                       a _) b _) ..) .. =>
    return match ← applyTransformations a, ← applyTransformations b with
    | some a', some b' =>
      mkApp2 (mkConst (Name.mkSimple "=")) a' b'
    | _      , _       => none
  | e                                   => pure e

end Smt.Bool
