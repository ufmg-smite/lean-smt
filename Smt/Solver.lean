/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Term

namespace Smt

inductive Kind where
  | cvc5
  | z3
  | vampire
  deriving Inhabited

instance : ToString Kind where
  toString := fun
    | .cvc5 => "cvc5"
    | .z3 => "z3"
    | .vampire => "vampire"

instance : Lean.KVMap.Value Kind where
  toDataValue k := toString k
  ofDataValue?
    | "cvc5"    => Kind.cvc5
    | "z3"      => Kind.z3
    | "vampire" => Kind.vampire
    | _         => none

register_option smt.solver.kind : Kind := {
  defValue := Kind.cvc5
  descr := "The solver to use for solving the SMT query."
}

register_option smt.solver.path : String := {
  defValue := "cvc5"
  descr := "The path to the solver used for solving the SMT query."
}

/-- The data-structure for SMT-LIB based solver. -/
structure Solver where
  commands : List String
  deriving Inhabited

namespace Solver

variable (solver : Solver)

/-- Set the SMT query logic to `l`. -/
def setLogic (l : String) : Solver :=
  ⟨s!"(set-logic {l})" :: solver.commands⟩

/-- Define a sort with name `id` and arity `n`. -/
def declareSort (id : String) (n : Nat) : Solver :=
  ⟨s!"(declare-sort {id} {Nat.toDigits 10 n})" :: solver.commands⟩

/-- Define a sort with name `id`. -/
def defineSort (id : String) (ss : List Term) (s : Term) : Solver :=
  ⟨(s!"(define-sort {id} ({paramsToString ss}) {s})") :: solver.commands⟩
  where
    paramsToString (ss) := String.intercalate " " (ss.map Term.toString)

/-- Declare a symbolic constant with name `id` and sort `s`. -/
def declareConst (id : String) (s : Term) : Solver :=
  ⟨s!"(declare-const {id} {s})" :: solver.commands⟩

/-- Define a constant with name `id` sort `s`, and value `v`. -/
def defineConst (id : String) (s : Term) (v : Term) : Solver :=
  ⟨s!"(define-const {id} {s} {v})" :: solver.commands⟩

/-- Declare an uninterpreted function with name `id` and sort `s`. -/
def declareFun (id : String) (s : Term) : Solver :=
  ⟨s!"(declare-fun {id} {s})" :: solver.commands⟩

/-- Define a function with name `id`, parameters `ps`, co-domain `s`,
    and body `t`. -/
def defineFun (id : String) (ps : List (String × Term)) (s : Term) (t : Term) :
  Solver :=
  ⟨s!"(define-fun {id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
  where
    paramsToString (ps : List (String × Term)) : String :=
      String.intercalate " " (ps.map (fun (n, s) => s!"({n} {s})"))

/-- Define a recursive function with name `id`, parameters `ps`, co-domain `s`,
    and body `t`. -/
def defineFunRec (id : String) (ps : List (String × Term)) (s : Term)
  (t : Term) :
  Solver :=
  ⟨s!"(define-fun-rec {id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
  where
    paramsToString (ps : List (String × Term)) : String :=
      String.intercalate " " (ps.map (fun (n, s) => s!"({n} {s})"))

/-- Assert that proposition `t` is true. -/
def assert (t : Term) : Solver :=
  ⟨s!"(assert {t})" :: solver.commands⟩

/-- Check if the query given so far is satisfiable and return the result. -/
def checkSat (kind : Kind) (path : String) : IO String := do
  let proc ← IO.Process.spawn {
    stdin := IO.Process.Stdio.piped
    stdout := IO.Process.Stdio.piped
    stderr := IO.Process.Stdio.piped
    cmd := path
    args := match kind with
      | Kind.cvc5    => #["-q"]
      | Kind.z3      => #["-in"]
      | Kind.vampire => #["--input_syntax", "smtlib2", "--output_mode", "smtcomp"]
  }
  let query := String.intercalate
                 "\n" ("(exit)\n" :: "(check-sat)" :: solver.commands).reverse
  proc.stdin.putStr query
  -- Close stdin to signal EOF to the solver.
  let (_, proc) ← proc.takeStdin
  _ ← proc.wait
  proc.stdout.readToEnd

end Solver
end Smt
