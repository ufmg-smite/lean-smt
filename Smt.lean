import Smt.General
import Smt.Nat
import Smt.Prop
import Smt.Solver
import Smt.String
import Smt.Tactic
import Smt.Term
