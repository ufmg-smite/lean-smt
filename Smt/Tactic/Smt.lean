/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean

import Smt.Dsl.Sexp
import Smt.Query
import Smt.Solver

namespace Smt

open Lean
open Lean.Elab
open Lean.Elab.Tactic

initialize
  registerTraceClass `smt.debug
  registerTraceClass `smt.debug.attr
  registerTraceClass `smt.debug.query
  registerTraceClass `smt.debug.transformer

/-- `smt` converts the current goal into an SMT query and checks if it is
satisfiable. By default, `smt` generates the minimum valid SMT query needed to
assert the current goal. However, that is not always enough:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt
```
For the theorem above, `smt` generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(check-sat)
```
which is missing the hypotheses `hp` and `hpq` required to prove the theorem. To
pass hypotheses to the solver, use `smt [h₁, h₂, ..., hₙ]` syntax:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt [hp, hpq]
```
The tactic then generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(declare-const p Bool)
(assert p)
(assert (=> p q))
(check-sat)
```
-/
syntax (name := smt) "smt" ("[" ident,* "]")? : tactic

def queryToString (commands : List String) : String :=
  String.intercalate "\n" ("(check-sat)\n" :: commands).reverse

def parseTactic : Syntax → TacticM (List Expr)
  | `(tactic| smt)              => pure []
  | `(tactic| smt [ $[$hs],* ]) => hs.toList.mapM (fun h => elabTerm h.raw none)
  | _                           => throwUnsupportedSyntax

@[tactic smt] def evalSmt : Tactic := fun stx => withMainContext do
  -- 1. Get the current main goal.
  let goalType ← Tactic.getMainTarget
  let goalId ← Lean.mkFreshMVarId
  Lean.Meta.withLocalDeclD goalId.name (mkNot goalType) fun g => do
  -- 2. Get the hints passed to the tactic.
  let mut hs ← parseTactic stx
  hs := hs.eraseDups
  -- 3. Generate the SMT query.
  let mut solver := Solver.mk []
  solver ← Query.generateQuery g hs solver
  let query := queryToString solver.commands
  -- 4. Run the solver.
  let kind := smt.solver.kind.get (← getOptions)
  let path := smt.solver.path.get? (← getOptions) |>.getD (toString kind)
  -- 5. Print the result.
  let res ← solver.checkSat kind path
  logInfo m!"goal: {goalType}\n\nquery:\n{query}\nresult: {res}"
  let out ← match Sexp.parse res with
    | .ok out => pure out
    | .error e => throwError "cannot parse solver output: {e}"
  if out.contains sexp!{sat} then
    throwError "unable to prove goal, it is false"
    -- TODO ask for countermodel and print
  else if !out.contains sexp!{unsat} then
    throwError "unable to prove goal"

end Smt
