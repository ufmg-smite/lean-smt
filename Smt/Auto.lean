/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Auto.Tactic
import Smt.Tactic.Smt

open Lean in
def smtSolverFunc (ls : Array Auto.Lemma) (is : Array Auto.Lemma) : MetaM Expr := do
  let fi l := do
    let userName ← mkFreshUserName `inst
    let type ← Meta.mkAppM ``Inhabited #[l.type]
    let value ← Meta.mkAppOptM ``Inhabited.mk #[l.type, l.proof]
    return { userName, type, value }
  let is ← is.mapM fi
  let fl l := do
    let userName ← mkFreshUserName `h
    let type ← Auto.Lam2D.approxSimpNF l.type
    let value := l.proof
    return { userName, type, value }
  let ls ← ls.mapM fl
  let h ← Meta.mkFreshExprMVar (Expr.const ``False [])
  let (_, mv) ← h.mvarId!.assertHypotheses is
  let (fvs, mv) ← mv.assertHypotheses ls
  mv.withContext do
    let hs := fvs.map (.fvar ·)
    _ ← Smt.smt mv hs.toList
    -- Note: auto should allow solvers to export new goals to users
    -- for mv in mvs do
    --   logInfo m!"new : {mv}"
    instantiateMVars h

attribute [rebind Auto.Native.solverFunc] smtSolverFunc
