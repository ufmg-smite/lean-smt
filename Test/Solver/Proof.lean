import Smt.Solver

open Smt Solver

def query : SolverM Sexp := do
  setLogic "QF_UF"
  assert (.symbolT "false")
  _ ← checkSat
  getProof

def main : IO Unit := do
  let ss ← createFromKind .cvc5 "cvc5" none
  let (res, ss) ← StateT.run query ss
  _ ← StateT.run exit ss
  println! "query:\n{Command.cmdsAsQuery ss.commands}\n\nres: unsat\n\nproof:\n{unquote (toString res)}"
where
  unquote s := s.extract ⟨8⟩ (s.endPos - ⟨2⟩)

#eval main
