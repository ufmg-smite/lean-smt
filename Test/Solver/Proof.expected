query:
(set-logic QF_UF)
(assert false)
(check-sat)
(get-proof)

res: unsat

proof:
import Smt.Reconstruction.Certifying

open Classical



set_option maxRecDepth 10000
set_option maxHeartbeats 500000


theorem th0 : False â†’ False :=
fun lean_a0 : False => by
show false from lean_a0



